viewing_party/party.py

```diff
@@ -1,23 +1,165 @@
+            genres_count[genre] = 1
+    
+    max_count = 0
+    most_frequently_watched_genre = ""
+
+    for genre in genres_count.keys():
+        if genres_count[genre] > max_count:
+            max_count = genres_count[genre]
+            most_frequently_watched_genre = genre
+    
+    return most_frequently_watched_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    unique_movies = []
+
+    for movie in user_data["watched"]:
+        is_unique = True
+        for friend in user_data["friends"]:
+            for friend_movie in friend["watched"]:
+                if movie["title"] == friend_movie["title"]:
+                    is_unique = False
+        
+        if is_unique:
+            unique_movies.append(movie)
+
+    return unique_movies
+
+
+def get_friends_unique_watched(user_data):
+    unique_movies = []
+
+    for friend in user_data["friends"]:
+        for friend_movie in friend["watched"]:
+            is_unique = True
+            for movie in user_data["watched"]:
+                if movie["title"] == friend_movie["title"]:
+                    is_unique = False
+        
+            if is_unique:
+                is_unique_and_new = True
+                for unique_movie in unique_movies:
+                    if unique_movie["title"] == friend_movie["title"]:
+                        is_unique_and_new = False
+
+                if is_unique_and_new:
+                    unique_movies.append(friend_movie)
+
+    return unique_movies
+
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    recommended_movies = []
+
+    friends_unique_watched = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_watched:
+        for subscription in user_data["subscriptions"]:
+            if subscription == movie["host"]:
+                recommended_movies.append(movie)
+
+    return recommended_movies
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    most_watched_genre_movies = []
+    
+    user_most_watched_genre = get_most_watched_genre(user_data)
+    friends_unique_watched_movies = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_watched_movies:
+        
+        if movie["genre"] == user_most_watched_genre:
+            most_watched_genre_movies.append(movie)
+
+    return most_watched_genre_movies
+
+
+def get_rec_from_favorites(user_data):
+    recommended_movies = []
+    unique_watched_user_movies = get_unique_watched(user_data)
+
+    for movie in unique_watched_user_movies:
+        for favorite in user_data["favorites"]:
+            if favorite["title"] ==  movie["title"]:

```

Comment:
Remember that instead of checking each title in `unique_watched_user_movies`, we can compare specific movies to each other.

I love the way you created `unique_watched_user_movies` using the helper function from wave 4. But now that we have that particular list, what if we just looped through the `user_data["favorites"]` and checked to see if each movie was in `unique_watched_user_movies`. This won't necessarily save us on complexity(remember that `in` will technically count as a loop in most cases), but it will save us a nested loop!

In the grand scheme of things, when we have a nested structure like the one in this project, it's a good idea to check entire dictionaries against each rather than check dictionary key-value pairs against each other!